Расширение функционала (изменение существующего пользователя). Добавление дополнительных параметров (uid, gid, home-dir)

spigell opened the pull request:
Здравствуйте. Мне нужно было немного расширить функционал этого плагина. В общем, я добавил:
 - указание gid и uid
 - Указание дополнительных групп
 - Указание домашнего каталога (также, есть параметр managehome `yes|no`, которые отвечает за создание/удаление/изменение домашнего каталога)
 - Смена логина

Кроме того, появллась downstream story *change*, которая призвана менять параметры у существующего пользователя (сделана через `usermod`)

Все на баше написано, да.

Посмотрите, как будет время.

spigell committed:
Added:
 - uid of user
 - gid of user
 - manage home dir of user

spigell committed:
fix README.md

spigell committed:
Remove set -x

spigell committed:
Added:
 - home directory
 - groups

spigell committed:
Added:
 - change script
Changed:
 - README

spigell committed:
Quick fix for README.md

spigell committed:
remove checks in story.check

melezhik commented:
Приветствую! Посмотрел немного. В целом норм, чуть позже более вдумчиво посмотрю код. Вопрос по существу. Зачем вам https://github.com/Spigell/user/blob/master/modules/change/story.bash#L66-L70 если вы все равно убрали из чеклиста проверку на то, что пользователь поменялся? IMHO я бы вообще бы обошелся без этого кода (66-70) ну и соответственно без функции check_old_user (https://github.com/Spigell/user/blob/master/modules/change/story.bash#L9) - мне кажется они здесь не нужны. Что думаете?

spigell commented:
Действительно, я убрал проверку, когда запускал плагин через sparrowdo. На некоторых хостах ничего не поменялось у меня и sparrow вылетал с ошибкой. Я забыл про такое поведение. Поэтому быстро убрал. 
Все-таки, я бы хотел, чтобы этот скрипт хоть что-то выводил. Пусть даже пока так в stdout одну фразу, потому что простая проверка по id как в create ничего не дает. Я так думаю. Все-таки изменение или неизменение в чем-то это важно, и надо их различать.

melezhik commented:
Ok, я вас я понял. Хорошо давайте тогда оставим эту часть кода.

Вот еще момент. https://github.com/Spigell/user/blob/master/hook.bash#L5-L12 Я бы хотел оставить hook.bash  в прежнем виде. Я обработку ситуауций пользователь "существует или нет"  я бы вынес в сами истории-модули (delete,create,change и так далее). Хочется оставить основной hook.bash максимально простым и не перегружать его логикой. 

И еще , вы используете https://github.com/Spigell/user/blob/master/hook.bash#L5 -
```
grep  $user /etc/passwd
```  
Вы уверенны, что такой проверка однозначна ? Например, она сработает и для  $user =  spa и для $user = sparrow при входных данных $name = spa  ... Или я возможно чего-то не понимаю ...

spigell commented:
Здравствуйте. Насчет `grep` вы абсолютно правы. Не доглядел. Возможно самый простой вариант через id, раз он часто используется. 
```
id $user &>/dev/null && is_user_exist=yes
if [[ $is_user_exist == yes ]]; then
...
```
Так мне больше нравится.
Насчет второго пункта, я сколько ни силился, не понимаю, чем здесь неправильный подход. В [package-generic](https://github.com/melezhik/package-generic/blob/master/hook.pl) примерно тоже самое. 
Просто эти условия нужны, чтобы для каждого действия скрипт уходил в правильную ветку. По умолчанию, переменная `action` равна `create`, поэтому плагин пойдет создавать. `useradd` выдаст, что такой пользователь создан и ничего делать не будет. Можно оставить как было до этого, но тогда в параметрах нужно четко указывать действие.

melezhik commented:
Давайте попробую сформулировать по-другому. Смотрите, что меня смущает в коде hook.bash плагина user:

```
user=$(config name)
action=$(config action)

if [[ `grep  $user /etc/passwd` ]]; then
  if [[ $action == delete ]]; then
    run_story delete
  else
    run_story change 
  fi
else
  run_story $action
fi
```

Здесь нет однозначного соответствия action -> история , это не страшно, но код становится слегка запутанным. В целом я даже не против, что бы в hook.bash было код, разбирающий входные параметры и делающий диспечерезацию ( в данном случае здесь речь идет о параметре action ), как это сделано у вас, только давайте сделаем это более явно, вот тогда как бы мне хотелось это сделать  ( далее привожу примерный код ) : 


 ```
user=$(config name)
action=$(config action)

  if [[ $action == delete ]] and [[ `grep  $user /etc/passwd` ]] ; then
    run_story delete
  elsif [ $action == delete ]
   set_stdout "user does not exist, skip this action" 
  fi

  if [[ $action == create ]] and [[  ! `grep  $user /etc/passwd` ]] ; then
    run_story create
  elsif [ $action == create ]
   set_stdout "user already exists, skip this action" 
  fi

  if [[ $action == change ]] and [[   `grep  $user /etc/passwd` ]] ; then
    run_story change
  elsif [ $action == change ]
   set_stdout "user does not exist, skip this action" 
  fi

```

( да, что бы set_stdout что-то напечатал в вывод нужно будет удалить файл meta.txt и создать пустой story.check )
 
Кстати в package-generic как раз такой подход и применен, в коде hook.pl есть четкий мэпинг между значением для action и вызываемой историей.


> Возможно самый простой вариант через id, раз он часто используется.

Согласен , id подойдет мне кажется вполне

melezhik commented:
Да и к предыдущему комментарию, приведенный мной пример bash кода, мне кажется, хорошо переписывается на `case ... esac`

spigell commented:
Я понял о чем вы говорите. Да, может мой вариант не совсем понятен будет. 

Однако мы кажется про разные вещи говорим. Я изменил код в hook.bash только для того, чтобы не указывать параметр `action`. плагин, исходя из того, что пользователь либо создан, либо нет - меняет или создает его соответственно. Параметр для удаления надо указывать уже четко.


Вот почему я это сделал:
На мой взгляд плагин user теряет некоторую долю смысла, если его не использовать вместе со sparrowdo (ну или через ssh, но это как то глупо). А раз sparrowdo это что-то похожее на cm, то хотелось бы, чтобы не надо было переписывать sparrowfile каждый раз. Вот если я напишу так:
```
task-run 'create user sparrow', 'user-spigell-dev', %(
  action   => 'create',
  name     => 'sparrow',
  home_dir => '/opt/sparrow',
  uid      => 453,
  gid      => 2300,
  groups   => 'tomatohelpers,wheel',
);

```
то на хостах, где есть этот юзер, он не придет к указанным параметрам, изменения проигнорируются. Просто потому что плагин пойдет по ветке create. Если я напишу в sparrowfile
`action => change`, то на тех хостах, где этого юзера нет, выскочит ошибка. Хотелось бы иметь одну точку входа в общем.

Я ни разу не разработчик, поясните, пожалуйста, что вы подразумеваете под словом мэпинг?

spigell commented:
```
user=$(config name)
action=$(config action)

id $user &>/dev/null && is_user_exist=yes

if [[ ! $is_user_exist == yes ]]; then
  echo "creating new user - $user"
  run_story create
fi

if [[ $is_user_exist == yes ]]; then
  echo "modifying existing user"
  run_story change
fi

if [[ $is_user_exist == yes ]] && [[ $action == delete ]]; then
  echo "delete user"
  run_story delete
fi
```
Может быть так правильней?

melezhik commented:
Я понял что вы пытались сделать изначально. Чуть позже отпишу

melezhik commented:
Да и ваш последний код мне нравится гораздо больше. Теперь все понятно ,что вы пытаетесь сделать. Давайте я ещё посмотрю на остальные части кода в merge request и напишу вам.

Мэппинг это такой слэнг разработчиков, от слова map , когда мы хотим поставить что-то в соответствие чему-то. Например мы пытаемся сопоставит значениям параметра action вызов соответсвующих скриптов ( модулей ).

spigell committed:
changed hook.bash - melezhik/user/#1

spigell committed:
Added:
 - user name in output
Fixes:
 - more accurate condition for change action
 - change modules without home_dir param works now
Misc
 - Added new_login to doc

melezhik commented on README.md:
```diff
@@ -8,23 +8,59 @@ Manages a linux user.
+
+## groups
+
+List of groups for user. Separated by comma.
```
List of user's groups. Should be comma separated list of items. For example:

    --param groups="developers,admins"

spigell replied:
Исправил.

spigell committed:
refs: melezhik/user#1
  - fix doc

melezhik commented:
вот еще соображения  данному MR.

Может быть есть смысл для _конечного пользователя_ просто оставить только  варианты значений для action = `("create"|"delete")` ? 

где  create в случае, когда пользователь существует, просто обновляет пользователя ( то, что у вас сейчас реализовано в модуле change/ - да и сам модуль нужно оставить, но просто не позволять вызывать его напрямую через action = change  )? 

Кстати в таком случае нет необходимости вводить  параметр `new_login` ( он будет просто браться из параметра  `name` )

Т.е. я просто за то, что бы не создавать лишние варианты для значений входных параметров без необходимости, т.е. в данном случае мы можем обойтись двумя:

 * action = create (  он же будет дефолтным, если значение action не указывать ) - который либо создает, либо обновляет пользователя с заданными параметрами
* action = delete - удаляет пользователя, если тот существует


Что думаете?

spigell commented:
Возможно так имели ввиду?
```
user=$(config name)
action=$(config action)

if [[ -z  $user ]];then
  echo "no user"
exit 1
fi

if [[ -z $action ]];then
  echo "I don't know what to do"
exit 1
fi

id $user &>/dev/null && is_user_exist=yes
if [[ ! $action == delete ]] && [[ ! $action == create ]] ;then
  set_stdout "Allowed action: <create|delete>"
fi

if [[ ! $is_user_exist == yes ]] && [[ $action == create ]]; then
  echo "creating new user - $user"
  run_story create
fi
if [[ $is_user_exist == yes ]] && [[ $action == create ]]; then
  echo "modifyng existing user"
  run_story change

fi
if [[ $is_user_exist == yes ]] && [[ $action == delete ]]; then
  echo "delete user"
  run_story delete
fi


```
С set_stdout разобрался.

--------

> Кстати в таком случае нет необходимости вводить параметр new_login ( он будет просто браться из параметра name )

Поясните этот момент. В этом плагине изменение пользователя меняем при помощи usermod. Нужно указать два параметра - старое имя и новое.

Spigell commented:
Все-таки есть вопрос по set_stdout. Почему строчка показывается даже если сценарий пошел не в ту ветку
```
[spigell@puppet_sparrow-test user]$ sudo strun --param action=delete --param name=test0-3  

[s] modules/delete/ at 2017-03-17 19:02:04
user test0-3 removed
ok      scenario succeeded
ok      text has 'removed'

[s] at 2017-03-17 19:02:04
Allowed action: <create|delete>
STATUS  SUCCEED

```

melezhik commented:
> Поясните этот момент. В этом плагине изменение пользователя меняем при помощи usermod. Нужно указать два параметра - старое имя и новое.

Да согласен, тогда отбой , принято.


> Возможно так имели ввиду? ...

Да, хотелось так ... так все чётко и понятно ... закомитье пожалуйста

melezhik commented:
Только видимо аналогичным образом нужно поменять "set_stdout "Allowed action: <create|delete>" на "echo 'Allowed action: <create|delete>'; exit 1" ?

Я потом ещё добавлю что бы в случае неуспешного статуса завершения хука выводился весь stdout, stderr , генерируемыйхуком, а не просто сообщение do_bash_hook failed, как сейчас. Тогда эти сообщения будут видны в отчете ...

spigell commented:
Я убрал проверку на параметры action и user так как она особо не нужна. `user` проверяется дальше. 
Вот у меня вопрос: Зачем нужна функция set_stdout(), если она требует story.check, что в свою очередь будет вызывать ошибку, так как будет отсуствовать story.bash? 
Я пока перетащил set_stdout в корневой story.bash
```
$ cat ./story.bash


action=$(config action)

if [[ ! $action == delete ]] && [[ ! $action == create ]] ;then
  echo "Allowed action: <create|delete>"
  exit 5
fi
exit 0

```

spigell commented:
> Только видимо аналогичным образом нужно поменять "set_stdout "Allowed action: <create|delete>" на "echo 'Allowed action: <create|delete>'; exit 1" ?

Но ведь `echo` в hook.bash не работает.

melezhik commented:
set_stdout() нужна что бы сгенерить вывод истории на уровне хука , если по каким то причинам не имеете или не хотите иметь сценарий для истории. И далее, если strun "понимает" что в хук генерит stdout, то он требует наличие проверочного файла , хотя бы пустого. Таким образом, вы не можете использовать set_stdout в мета историях, так как у мета истории по определению нет сценариев и нет связанных с ними проверочных файлов , мета истории - это просто истории-контейнеры для вызова других историй модулей. Поэтому, в случае с корневым hook.bash, если мы хотим оставить корневую историю как это и было мета историей , необходимо убрать все вызовы set_stdout

melezhik commented:
> Но ведь echo в hook.bash не работает.

Сработает, просто вы не увидите вывод от echo в отчете, но stun все равно запишет этот вывод в отдельный файл , название ( путь ) которого, в случае ненулевого статуса завершения хука будет выведенного в отчёт, с тем что пользователь мог потом этот файл прочесть для более подробной информации. Но как я уже сказал, я планирую вывод stdout, stderr от хук файла в случае ошибки выводить сразу в отчёт для наглядности.

melezhik commented:
Да и очень важно понимать что set_stdout и echo в хуке - совершенно разнеые вещи , хотя и то и другое используется для для генерации строчек вывода stdout  , только в разных контекстах, один для вывода stdout  истории другой для вывода stdout от хука. Первый требует наличия проверочного файла , второй не требует .

spigell committed:
refs: melezhik/user#1
 - added restriction for action. Only create|delete allowed.
 - removed "user" in *.check files.
 - changed docs.

spigell commented:
Закоммитил. Сделал как вы сказали выше. Так меньше файлов получается и более логично.

melezhik commented:
Ага отлично, чуть позже гляну , спасибо . Думаю еже скоро смерджу и если что какие то мелочевки уже сам доведу до ума ...

melezhik commented:
Спасибо! 
Замерджил, сделаю еще небольшие корректировки и зарелюзю.

melezhik commented:
зарелизил в версии 0.1.0 - https://sparrowhub.org/info/user , если будет время - просьба потестировать - изменений было достаточно много ... , если что-то захотите менять - делайте новый форк - снова создавайте PR

spigell commented:
Хорошо. На днях буду менять свои sparrowfiles. Там и потестирую.

melezhik commented:
да, спасибо, если что - дайте знать. Если будет время, то и сам постараюсь добавить тесты
